### PR TITLE
Cu jm add new item

### LINKED ISSUE #4: 4. As a user, I want to add a new item to my shopping list so I can start recording purchases.
A shopping list item consists of the following data points:

* Name of item
* How soon are you likely to buy it again?
  * Soon (in the next 7 days)
  * Kind of soon (in the next 14 days)
  * Not soon (in the next 30 days)
* Last purchased date

AC:

* User is presented with a form that allows them to enter the name of the item and select how soon they anticipate needing to buy it again (Soon, Kind of Soon, Not Soon)
* When the user submits the form, the item is saved to the database, associated with the user’s token
* Along with the item name, an integer corresponding to the estimated number of days until next purchase is saved: `7` for “Soon”, `14` for “Kind of Soon”, and `30` for “Not Soon”
* The last purchased date should be set to `null` initially (i.e. you can create an item without purchasing it)
* Item names should be displayed as a simple, unordered list on the list view after they’re added

### PR DESCRIPTION
**Pull request by Amaka202**
## Description

This PR adds two components: one for adding new items to the database associated with the user's token and another for displaying the list of items as an unordered list.

## Related Issue

Closes #4 

## Acceptance Criteria

- [x] User is presented with a form that allows them to enter the name of the item and select how soon they anticipate needing to buy it again (Soon, Kind of Soon, Not Soon)
- [x] When the user submits the form, the item is saved to the database, associated with the user’s token
- [x] Along with the item name, an integer corresponding to the estimated number of days until next purchase is saved: 7 for “Soon”, 14 for “Kind of Soon”, and 30 for “Not Soon”
- [x] The last purchased date should be set to null initially (i.e. you can create an item without purchasing it)
- [x] Item names should be displayed as a simple, unordered list on the list view after they’re added

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

Screenshot (49)
Screenshot (48)

## Testing Steps / QA Criteria

- From your terminal, pull down this branch with git pull origin cu-jm-add-new-item and check that branch out with git checkout cu-jm-add-new-item
- npm install to install app dependencies
- Fill out the form with a new item, choose an option among the radio buttons and submit
- Navigate to the lists view to view the added list

### CODE REVIEW

File: src/components/AddItems.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+  const shoppingListItemNameHandler = (event) => {
+    setShoppingListItemName(event.target.value);
+  };
+  const daysLeftForNextPurchaseHandler = (event) => {
+    setDaysLeftForNextPurchase(parseInt(event.target.value));
+  };
+  function submitShoppingListItemHandler(event) {
+    event.preventDefault();
+
+    localStorage.setItem('token', 288838881888283);
```
Comment (rb50) [inline src/components/AddItems.js:21]: Am I right that when we merge both PRs we can delete this line, as the real token should be set by our PR?

File: src/components/AddItems.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+      shoppingListItemName,
+      daysLeftForNextPurchase,
+      lastPurchasedOn: null,
+    };
+
+    db.where('token', '==', userToken)
+      .get()
+      .then((data) => {
+        if (data.docs.length) {
+          console.log(data.empty);
```
Comment (rb50) [inline src/components/AddItems.js:40]: I think the console.log can be deleted?

**Review by rb50 — COMMENTED** [review]
This is great! It's really clear code and easy to read. I just had a couple of questions and comments which I've added, mainly about agreeing our naming of things. Also just to mention that in your screenshots in the PR it shows an ordered list, but in the branch when I run it it shows the unordered list (as per the acceptance criteria) so I think the screenshots just need updating.

File: src/App.js [hunk 3]
```diff
@@ -1,26 +1,21 @@
...
-import ItemsList from './components/ItemsList';
-import AddItemsToList from './components/AddItemsToList';
 
 function App() {
   return (
     <div className="App">
       <Switch>
         <Route path="/list">
-          <List />
+          <ItemsList />
```
[the remark this answers is not in the document]
Comment (Amaka202) [reply src/App.js:13]: Ok, go ahead and remove it from your end. So that we will have just "itemsList" and "AddItems"

File: src/components/AddItems.js (hunk 1 above)
Comment (Amaka202) [reply src/components/AddItems.js:21]: Exactly, that line is just to be used to test our functionality. it will be taken out before merging issue 4.

File: src/components/AddItems.js [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+  };
+  const daysLeftForNextPurchaseHandler = (event) => {
+    setDaysLeftForNextPurchase(parseInt(event.target.value));
+  };
+  function submitShoppingListItemHandler(event) {
+    event.preventDefault();
+
+    localStorage.setItem('token', 288838881888283);
+
+    const userToken = localStorage.getItem('token');
```
[the remark this answers is not in the document]
Comment (Amaka202) [reply src/components/AddItems.js:23]: ok Great!

File: src/components/AddItems.js (hunk 2 above)
Comment (Amaka202) [reply src/components/AddItems.js:40]: done I will update the PR asap.

File: src/components/AddItems.js (hunk 1 above)
Comment (saliainred) [reply src/components/AddItems.js:21]: This code looks so well structured and readable, great work !

**Review by saliainred — COMMENTED** [review]
This code looks so well structured and readable. Great work. Glad that everything works fine now.

File: src/components/ItemsList.js [hunk 5]
```diff
@@ -1,32 +1,47 @@
...
+                  <p>{val}</p>
+                </li>
+              );
+            })
+          ) : (
+            <p>Loading ...</p>
+          )}
+        </ul>
+      </div>
     </div>
```
Comment (Lukabaramishvili) [inline src/components/ItemsList.js:45]: I love how you check if there are items in the array and then render each individual item. 

I have one suggestion:
Right now if the `shopping_list` collection is empty we always render `<p>Loading...</p>`. I think it would be better to 
1. First check if we are fetching the data from the database. While it is fetching render `Loading...`. (if loading is true, render `Loading...`).
2. When the data is fetched, show the items.
3. Otherwise you can show either an empty page or render `List is empty`. 

In other words, what I'm trying to suggest is very similar to what you had before: 

Screen Shot 2021-01-16 at 3 59 52 PM

File: src/components/AddItems.js [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+    if (shoppingListItemName === '') {
+      alert('Please enter item name...');
+      return;
+    }
```
Comment (Lukabaramishvili) [inline src/components/AddItems.js:28]: I love that you guard for an empty string here. Very thoughtful 👏

File: src/components/AddItems.js [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        } else {
+          db.add({
+            token: userToken,
+            items: [values],
+          });
+          setShoppingListItemName('');
+          setDaysLeftForNextPurchase(7);
+        }
+      });
+  }
```
Comment (Lukabaramishvili) [inline src/components/AddItems.js:54]: Working well with the firebase commands here 👍 🚀

File: src/components/AddItems.js [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+      <div>
+        <label>
+          Name
+          <input
+            type="text"
+            value={shoppingListItemName}
+            onChange={shoppingListItemNameHandler}
+          />
+        </label>
+      </div>
```
Comment (Lukabaramishvili) [inline src/components/AddItems.js:67]: This form is super clean and well organized. Nice work 👏

File: src/components/AddItems.js (hunk 1 above)
Comment (Lukabaramishvili) [reply src/components/AddItems.js:21]: Yes yes, I totally agree Julia. Variables are named so well. It reads very well and is self-explanatory.

**Review by Lukabaramishvili — COMMENTED** [review]
Great work team, 👏 I had some minor suggestions and comments. The main one is the loading state in the `AddItem` component. 

More details below. Thank you!

**Review by faysvas — COMMENTED** [review]
Great work! I made a small suggestion but it's not necessary to follow it.
Luka already did a great code review so I don't have much more to say :100:

File: src/components/AddItems.js (hunk 1 above)
Comment (faysvas) [reply src/components/AddItems.js:21]: An additional note: I see that the other team is using 'newToken' as their token name so make sure to coordinate with them before merging :)

File: src/components/ItemsList.js (hunk 5 above)
Comment (Amaka202) [reply src/components/ItemsList.js]: ok. I have updated it to not show loading at all.

File: src/components/ItemsList.js [hunk 9]
```diff
@@ -1,32 +1,44 @@
...
+          let itemsList = data.docs[0]
+            .data()
+            .items.map((element) => element.shoppingListItemName);
+          if (mounted) {
+            setItems(itemsList);
+          }
+        }
+      });
+    return () => (mounted = false);
+  });
```
Comment (faysvas) [inline src/components/ItemsList.js:26]: This code may be causing the quota issue. It runs in a loop

Comment (salmanmanekia) [reply src/components/ItemsList.js:26]: does giving [items] as the second argument of the useEffect help ?

Comment (salmanmanekia) [reply src/components/ItemsList.js:26]: ..or just remove the useEffect completely and the call to the firestore happens only once :)

**Review by rb50 — APPROVED** [review]

**Review by alexlsalt — COMMENTED** [review]
Great work here! Looking forward to seeing the merged product later tonight during our weekly sync 🙌

**Review by salmanmanekia — APPROVED** [review]
